More configurable path

user202729 opened the pull request:
`cf parse`, `cf submit` and `cf clone` should work fine now. The rest should be unaffected.

----------

Should fix #67.

I think the current `cf clone` is incorrect (the directory structure generated cannot be parsed by `cf submit`) (not tested)

I don't know if there's any part of the code that modifies the other parts of the info. (in that case the path is no longer correct, and it's necessary to store the configuration option in the info object to dynamically compute the path when needed)

`%` is used to escape the patterns. To have a `%`, enter two percents. Path separator is always slash, even on Windows. The path specifier is listed in decreasing priority.

The default-generated configuration is the same as the >=0.9.0 behavior. Pre-0.9.0 behavior can be set by something like this

```json
    [
      "contest",
      "%contestID%/%problemID%"
    ],
    [
      "gym",
      "%contestID%/%problemID%"
    ],
    [
      "group",
      "group_%groupID%/%contestID%/%problemID%"
    ],
    [
      "acmsguru",
      "acmsguru/%problemID%"
    ]
```

user202729 committed:
More configurable path

`cf clone` is incomplete.

user202729 committed:
Bug fix

gnojus commented:
Maybe use `map` instead off array in this configuration?
I'm also thinking that we should support comments in config, with something like this: https://github.com/muhammadmuzzammil1998/jsonc.

user202729 commented:
> Maybe use `map` instead off array in this configuration?

->

> The path specifier is listed in decreasing priority.

gnojus commented:
Oh. Then I would still prefer something like
```json
[
    {
      "type": "contest",
      "path": "%contestID%/%problemID%"
    },
    {
      "type": "gym",
      "path": "%contestID%/%problemID%"
    },
    {
      "type": "group",
      "path": "group_%groupID%/%contestID%/%problemID%"
    },
    {
      "type": "acmsguru",
      "path": "acmsguru/%problemID%"
    }
]
```
Just to avoid the inner arrays

user202729 committed:
Fix another bug (typo)

user202729 commented:
The change is actually pretty intrusive -- with the other code unchanged, it's not possible to configure the problem directory path -- so for example it would not be possible to use `%contestID%_%problemID%`. Since I didn't change the "other code" now it would parse to (contest)/(problem)/(problem)/ directory instead.

user202729 committed:
Use a struct for path specifier instead of array

Note that this change is config-incompatible with the previous commit, so it may be necessary
to hand-edit the config file.

user202729 committed:
Refactor code: merge ArgRegStr and ArgType to make code more readable

user202729 committed:
Better detection of problem type

Path-based detection is not completely reliable because the user may set patterns arbitrarily.

user202729 committed:
Move Langs and LangsExt from client to util to avoid circular dependency

user202729 committed:
Fix cf parse

user202729 committed:
Bug fix for problem type detection

user202729 committed:
Raise error on invalid specifier

user202729 committed:
cf clone should put all files in handle/ folder

user202729 committed:
Comment cleanup
